File: 15829/15829.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    15829.py                                           :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2024/03/10 14:33:36 by jihoolee          #+#    #+#              #
#    Updated: 2024/03/10 14:36:18 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

R: int = 31
M: int = 1234567891


def hash(string: str):
    result = 0

    for idx, c in enumerate(string):
        result += (ord(c) - ord('a') + 1) * (R ** idx)
    return result % M


def main() -> None:
    _ = input()
    string = input()

    print(hash(string))


if __name__ == '__main__':
    main()

File: 15829/15829.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   15829.cpp                                          :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2024/03/10 13:56:45 by jihoolee          #+#    #+#             */
/*   Updated: 2024/03/10 14:33:07 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <string>

const int r = 31;
const int64_t M = 1234567891;

int64_t hash(std::string str) {
  int64_t result = 0;
  int64_t acc_r = 1;

  for (size_t idx = 0; idx < str.length(); ++idx) {
    result = (result + (str[idx] - 'a' + 1) * acc_r) % M;
    acc_r =  (acc_r * r) % M;
  }
  return result % M;
}

int main(void) {
  int         length;
  std::string str;

  std::cin >> length;
  (void)length;
  std::cin >> str;
  std::cout << hash(str);
  return 0;
}
